File: cloudsdk/src/main/java/io/particle/android/sdk/cloud/Responses.java
package io.particle.android.sdk.cloud;

import com.google.gson.annotations.SerializedName;

import java.util.Date;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import io.particle.android.sdk.cloud.Responses.Models.CoreInfo;
import io.particle.android.sdk.cloud.models.ParticleNetworkData;

/**
 * All API responses, collected together in one outer class for simplicity's sake.
 */
public class Responses {

    /**
     * ...and to go with the responses, a series of model objects only
     * used internally when dealing with the REST API, never returned
     * outside of the cloudapi package.
     */
    public static class Models {


        public static class CoreInfo {

            @SerializedName("last_app")
            public final String lastApp;

            @SerializedName("last_heard")
            public final Date lastHeard;

            public final boolean connected;

            public final String deviceId;

            public CoreInfo(String lastApp, Date lastHeard, boolean connected, String deviceId) {
                this.lastApp = lastApp;
                this.lastHeard = lastHeard;
                this.connected = connected;
                this.deviceId = deviceId;
            }
        }


        /**
         * Represents a single Particle device in the list returned
         * by a call to "GET /v1/devices"
         */
        public static class SimpleDevice {

            public final String id;

            public final String name;

            public final boolean cellular;

            public final String imei;

            @SerializedName("last_iccid")
            public final String lastIccid;

            @SerializedName("current_build_target")
            public final String currentBuild;

            @SerializedName("default_build_target")
            public final String defaultBuild;

            @SerializedName("connected")
            public final boolean isConnected;

            @SerializedName("product_id")
            public final int productId;

            @SerializedName("platform_id")
            public final int platformId;

            @SerializedName("last_ip_address")
            public final String ipAddress;

            @SerializedName("status")
            public final String status;

            @SerializedName("last_heard")
            public final Date lastHeard;

            public SimpleDevice(String id, String name, boolean isConnected, boolean cellular,
                                String imei, String lastIccid, String currentBuild, String defaultBuild, int platformId,
                                int productId, String ipAddress, String status, Date lastHeard) {
                this.id = id;
                this.name = name;
                this.isConnected = isConnected;
                this.cellular = cellular;
                this.imei = imei;
                this.lastIccid = lastIccid;
                this.currentBuild = currentBuild;
                this.defaultBuild = defaultBuild;
                this.platformId = platformId;
                this.productId = productId;
                this.ipAddress = ipAddress;
                this.status = status;
                this.lastHeard = lastHeard;
            }
        }

        /**
         * Represents a single Particle device as returned from the
         * call to "GET /v1/devices/{device id}"
         */
        public class CompleteDevice {
            @SerializedName("id")
            public final String deviceId;

            public final String name;

            public final boolean cellular;

            public final String imei;

            @SerializedName("last_iccid")
            public final String lastIccid;

            @SerializedName("current_build_target")
            public final String currentBuild;

            @SerializedName("default_build_target")
            public final String defaultBuild;

            @SerializedName("connected")
            public final boolean isConnected;

            public final Map<String, String> variables;

            public final List<String> functions;

            @SerializedName("cc3000_patch_version")
            public final String version;

            @SerializedName("product_id")
            public final int productId;

            @SerializedName("platform_id")
            public final int platformId;

            @SerializedName("last_ip_address")
            public final String ipAddress;

            @SerializedName("last_app")
            public final String lastAppName;

            @SerializedName("status")
            public final String status;

            @SerializedName("device_needs_update")
            public final boolean requiresUpdate;

            @SerializedName("last_heard")
            public final Date lastHeard;

            @SerializedName("serial_number")
            public final String serialNumber;

            @SerializedName("mobile_secret")
            public final String mobileSecret;

            CompleteDevice(String deviceId, String name, boolean isConnected, boolean cellular,
                           String imei, String lastIccid, String currentBuild, String defaultBuild,
                           Map<String, String> variables, List<String> functions, String version,
                           int productId, int platformId, String ipAddress, String lastAppName,
                           String status, boolean requiresUpdate, Date lastHeard,
                           String serialNumber, String mobileSecret) {
                this.deviceId = deviceId;
                this.name = name;
                this.isConnected = isConnected;
                this.cellular = cellular;
                this.imei = imei;
                this.lastIccid = lastIccid;
                this.currentBuild = currentBuild;
                this.defaultBuild = defaultBuild;
                this.variables = variables;
                this.functions = functions;
                this.version = version;
                this.productId = productId;
                this.platformId = platformId;
                this.ipAddress = ipAddress;
                this.lastAppName = lastAppName;
                this.status = status;
                this.requiresUpdate = requiresUpdate;
                this.lastHeard = lastHeard;
                this.serialNumber = serialNumber;
                this.mobileSecret = mobileSecret;
            }
        }

    }


    public static class TokenResponse {

        public final String token;

        public TokenResponse(String token) {
            this.token = token;
        }
    }


    public static class CallFunctionResponse {

        @SerializedName("id")
        public final String deviceId;

        @SerializedName("name")
        public final String deviceName;

        public final boolean connected;

        @SerializedName("return_value")
        public final int returnValue;

        public CallFunctionResponse(String deviceId, String deviceName, boolean connected,
                                    int returnValue) {
            this.deviceId = deviceId;
            this.deviceName = deviceName;
            this.connected = connected;
            this.returnValue = returnValue;
        }
    }


    public static class LogInResponse {

        @SerializedName("expires_in")
        public final long expiresInSeconds;

        @SerializedName("access_token")
        public final String accessToken;

        @SerializedName("refresh_token")
        public final String refreshToken;

        @SerializedName("token_type")
        public final String tokenType;

        public LogInResponse(long expiresInSeconds, String accessToken, String refreshToken, String tokenType) {
            this.expiresInSeconds = expiresInSeconds;
            this.accessToken = accessToken;
            this.refreshToken = refreshToken;
            this.tokenType = tokenType;
        }
    }


    public static class SimpleResponse {

        public final boolean ok;
        public final String error;

        public SimpleResponse(boolean ok, String error) {
            this.ok = ok;
            this.error = error;
        }

        @Override
        public String toString() {
            return "SimpleResponse [ok=" + ok + ", error=" + error + "]";
        }
    }


    public static class ClaimCodeResponse {

        @SerializedName("claim_code")
        public final String claimCode;

        @SerializedName("device_ids")
        public final String[] deviceIds;

        public ClaimCodeResponse(String claimCode, String[] deviceIds) {
            this.claimCode = claimCode;
            this.deviceIds = deviceIds;
        }
    }

    public abstract static class ReadVariableResponse<T> {

        @SerializedName("cmd")
        public final String commandName;

        @SerializedName("name")
        public final String variableName;

        public final T result;

        public final Models.CoreInfo coreInfo;

        public ReadVariableResponse(String commandName, String variableName,
                                    Models.CoreInfo coreInfo, T result) {
            this.commandName = commandName;
            this.variableName = variableName;
            this.result = result;
            this.coreInfo = coreInfo;
        }
    }


    public static class FirmwareUpdateInfoResponse {

        @SerializedName("binary_url")
        public final String nextFileUrl;

        public FirmwareUpdateInfoResponse(String nextFileUrl) {
            this.nextFileUrl = nextFileUrl;
        }
    }


    public static class ReadIntVariableResponse extends ReadVariableResponse<Integer> {

        public ReadIntVariableResponse(String commandName, String variableName, CoreInfo coreInfo,
                                       Integer result) {
            super(commandName, variableName, coreInfo, result);
        }
    }


    public static class ReadDoubleVariableResponse extends ReadVariableResponse<Double> {

        public ReadDoubleVariableResponse(String commandName, String variableName, CoreInfo coreInfo,
                                          Double result) {
            super(commandName, variableName, coreInfo, result);
        }
    }


    public static class ReadStringVariableResponse extends ReadVariableResponse<String> {

        public ReadStringVariableResponse(String commandName, String variableName, CoreInfo coreInfo,
                                          String result) {
            super(commandName, variableName, coreInfo, result);
        }
    }


    public static class ReadObjectVariableResponse extends ReadVariableResponse<Object> {

        public ReadObjectVariableResponse(String commandName, String variableName, CoreInfo coreInfo,
                                          Object result) {
            super(commandName, variableName, coreInfo, result);
        }
    }


    public abstract static class PagedDataResponse {

        public static class PagedMetadata {
            public final int totalRecords;
            public final int totalPages;

            public PagedMetadata(int totalRecords, int totalPages) {
                this.totalRecords = totalRecords;
                this.totalPages = totalPages;
            }
        }


        @NonNull
        public final PagedMetadata meta;

        protected PagedDataResponse(@NonNull PagedMetadata meta) {
            this.meta = meta;
        }

    }


    public static class MeshNetworksResponse extends PagedDataResponse {

        @NonNull
        public final List<ParticleNetworkData> networks;

        protected MeshNetworksResponse(
                @NonNull PagedMetadata meta,
                @NonNull List<ParticleNetworkData> networks) {
            super(meta);
            this.networks = networks;
        }
    }


    public static class MeshNetworkMembershipsResponse extends PagedDataResponse {

        public static class MeshMembership {

            @NonNull
            public final String id;
            public final boolean gateway;

            public MeshMembership(String id, boolean gateway) {
                this.id = id;
                this.gateway = gateway;
            }
        }


        public static class MeshNetworkDeviceMemberships {
            @NonNull
            public final String id;
            @NonNull
            @SerializedName("network")
            public final MeshMembership membership;

            public MeshNetworkDeviceMemberships(String id, MeshMembership membership) {
                this.id = id;
                this.membership = membership;
            }
        }

        @NonNull
        @SerializedName("devices")
        public final List<MeshNetworkDeviceMemberships> memberships;

        public MeshNetworkMembershipsResponse(
                @NonNull PagedMetadata meta,
                @NonNull List<MeshNetworkDeviceMemberships> memberships
        ) {
            super(meta);
            this.memberships = memberships;
        }
    }


    public static class MeshNetworkRegistrationResponse {

        public static class RegisteredNetwork {
            @NonNull
            public final String id;
            @NonNull
            public final String name;
            @NonNull
            public final ParticleNetworkState state;
            @NonNull
            public final ParticleNetworkType type;

            public RegisteredNetwork(String id, String name, ParticleNetworkState state,
                                     ParticleNetworkType type) {
                this.id = id;
                this.name = name;
                this.state = state;
                this.type = type;
            }
        }

        @NonNull
        public final RegisteredNetwork network;

        public MeshNetworkRegistrationResponse(@NonNull RegisteredNetwork network) {
            this.network = network;
        }

    }


}

File: cloudsdk/src/main/java/io/particle/android/sdk/cloud/Responses.kt
package io.particle.android.sdk.cloud

import com.google.gson.annotations.SerializedName
import io.particle.android.sdk.cloud.Responses.Models.CoreInfo
import io.particle.android.sdk.cloud.models.ParticleNetworkData
import java.util.*

/**
 * All API responses, collected together in one outer class for simplicity's sake.
 */
class Responses {

    /**
     * ...and to go with the responses, a series of model objects only
     * used internally when dealing with the REST API, never returned
     * outside of the cloudapi package.
     */
    class Models {

        class CoreInfo(
            @SerializedName("last_app")
            val lastApp: String,
            @SerializedName("last_heard")
            val lastHeard: Date,
            val connected: Boolean,
            val deviceId: String
        )

        /**
         * Represents a Particle device from "GET /v1/devices/{device id}"
         * and "GET /v1/devices"
         */
        class CompleteDevice internal constructor(
            @SerializedName("id")
            val deviceId: String,
            val name: String,
            @SerializedName("connected")
            val isConnected: Boolean,
            val cellular: Boolean,
            val imei: String,
            val iccid: String,
            @SerializedName("last_iccid")
            val lastIccid: String,
            @SerializedName("current_build_target")
            val currentBuild: String,
            @SerializedName("default_build_target")
            val defaultBuild: String,
            val variables: Map<String, String>?,
            val functions: List<String?>?,
            @SerializedName("product_id")
            val productId: Int,
            @SerializedName("platform_id")
            val platformId: Int,
            @SerializedName("last_ip_address")
            val ipAddress: String,
            @SerializedName("last_app")
            val lastAppName: String,
            @SerializedName("status")
            val status: String,
            @SerializedName("last_heard")
            val lastHeard: Date,
            @SerializedName("serial_number")
            val serialNumber: String,
            @SerializedName("mobile_secret")
            val mobileSecret: String,
            @SerializedName("system_firmware_version")
            val systemFirmwareVersion: String,
            val notes: String?
        )

    }


    class TokenResponse(val token: String)


    class CallFunctionResponse(
        @SerializedName("id")
        val deviceId: String,
        @SerializedName("name")
        val deviceName: String,
        val connected: Boolean,
        @SerializedName("return_value")
        val returnValue: Int
    )


    class LogInResponse(
        @SerializedName("expires_in")
        val expiresInSeconds: Long,
        @SerializedName("access_token")
        val accessToken: String,
        @SerializedName("refresh_token")
        val refreshToken: String,
        @SerializedName("token_type")
        val tokenType: String
    )


    class SimpleResponse(val ok: Boolean, val error: String) {

        override fun toString(): String {
            return "SimpleResponse [ok=$ok, error=$error]"
        }
    }


    class ClaimCodeResponse(
        @SerializedName("claim_code")
        val claimCode: String,
        @SerializedName("device_ids")
        val deviceIds: Array<String>
    )

    class FirmwareUpdateInfoResponse(
        @SerializedName("binary_url")
        val nextFileUrl: String
    )

    abstract class ReadVariableResponse<T>(
        @SerializedName("cmd")
        val commandName: String,
        @SerializedName("name")
        val variableName: String,
        val coreInfo: Models.CoreInfo,
        val result: T
    )

    class ReadIntVariableResponse(
        commandName: String,
        variableName: String,
        coreInfo: CoreInfo,
        result: Int
    ) : ReadVariableResponse<Int>(commandName, variableName, coreInfo, result)


    class ReadDoubleVariableResponse(
        commandName: String,
        variableName: String,
        coreInfo: CoreInfo,
        result: Double
    ) : ReadVariableResponse<Double>(commandName, variableName, coreInfo, result)


    class ReadStringVariableResponse(
        commandName: String,
        variableName: String,
        coreInfo: CoreInfo,
        result: String
    ) : ReadVariableResponse<String>(commandName, variableName, coreInfo, result)


    class ReadObjectVariableResponse(
        commandName: String,
        variableName: String,
        coreInfo: CoreInfo,
        result: Any?
    ) : ReadVariableResponse<Any?>(commandName, variableName, coreInfo, result)


    class MeshNetworkRegistrationResponse(val network: RegisteredNetwork) {

        data class RegisteredNetwork(
            val id: String,
            val name: String,
            val state: ParticleNetworkState,
            val type: ParticleNetworkType
        )

    }

    data class DeviceMeshMembership(
        @SerializedName("id")
        val deviceId: String,
        @SerializedName("network")
        val membership: MeshMembership
    )

    data class MeshMembership(
        @SerializedName("id")
        val networkId: String,
        @SerializedName("role")
        val roleData: MeshRoleData
    )

    data class MeshRoleData(
        @SerializedName("gateway")
        val isGateway: Boolean
    )


    class PingResponse(val online: Boolean)


    data class CardOnFileResponse(
        val card: CardOnFile?
    ) {

        data class CardOnFile(

            val last4: String?,

            val brand: String?,

            @SerializedName("exp_month")
            val expiryMonthString: String?,

            @SerializedName("exp_year")
            val expiryYearString: String?,

            @SerializedName("id")
            val token: String?
        )

    }

}
